### PR TITLE
Reuse IMessageSession and set PrefetchCount for Sessions. Improve Tri…

### LINKED ISSUE #88: Reuse IMessageSession and set PrefetchCount for Sessions

### PR DESCRIPTION
**Pull request by alrod**
…gger Details. Fixes #88, #85.

### CODE REVIEW

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs [hunk 1]
```diff
@@ -251,11 +249,12 @@ internal void StartMessageBatchReceiver(CancellationToken cancellationToken)
...
-                        if (_isSessionsEnabled)
+                        if (_isSessionsEnabled && ( receiver == null || receiver.IsClosedOrClosing))
                         {
                             try
                             {
                                 receiver = await sessionClient.AcceptMessageSessionAsync();
+                                receiver.PrefetchCount = _serviceBusOptions.PrefetchCount;
                             }
                             catch (ServiceBusTimeoutException)
                             {
```
Comment (sidkri) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs:260]: Would it be useful to have a debug statement here when we there are not sessions to accept or too much of a perf. hit?

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs [hunk 2]
```diff
@@ -302,11 +301,13 @@ internal void StartMessageBatchReceiver(CancellationToken cancellationToken)
                                     }
                                 }
                             }
-
-                            // Close the session and release the session lock
-                            if (_isSessionsEnabled)
+                            else
                             {
-                                await receiver.CloseAsync();
+                                // Close the session and release the session lock
```
Comment (sidkri) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs:306]: Would be good to enhance this comment and explain that we are only closing the session after draining all prefetched messages.

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs [hunk 3]
```diff
@@ -302,11 +301,13 @@ internal void StartMessageBatchReceiver(CancellationToken cancellationToken)
                                     }
                                 }
                             }
-
-                            // Close the session and release the session lock
-                            if (_isSessionsEnabled)
+                            else
```
Comment (nemakam) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs:304]: >else [](start = 28, length = 4)

Should this else come after another `}`? Or I might be counting wrong. I assume the intention is to close the receiver only when `messages == null`

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs [hunk 4]
```diff
@@ -251,11 +249,12 @@ internal void StartMessageBatchReceiver(CancellationToken cancellationToken)
                             return;
                         }
 
-                        if (_isSessionsEnabled)
+                        if (_isSessionsEnabled && ( receiver == null || receiver.IsClosedOrClosing))
                         {
                             try
                             {
                                 receiver = await sessionClient.AcceptMessageSessionAsync();
+                                receiver.PrefetchCount = _serviceBusOptions.PrefetchCount;
```
Comment (nemakam) [inline src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs:257]: There are some rare cases where this returns null. It was a bug long time ago, not sure if its fixed or not. For the sake of precautions, I would put a null check after this and if this happens to be null, do a `continue`

**Comment by nemakam** [comment]
>                                     if (_isSessionsEnabled)

Not related to you PR, but why are we not abandoning the messages in case of non-sessions?

---
Refers to: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs:293 in 9799ed1. [](commit_id = 9799ed1b7355c4fd027ff59690932c6d62a147b3, deletion_comment = False)

**Comment by nemakam** [comment]
>                                 }

If this throws, you will never end up closing and recreating the session. Lets say you enter into SessionLockLost exception once, and then you'd never recovver from that

---
Refers to: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs:287 in 9799ed1. [](commit_id = 9799ed1b7355c4fd027ff59690932c6d62a147b3, deletion_comment = False)

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs (hunk 3 above)
Comment (sidkri) [reply src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs:304]: Good catch, I just counted and it does need to be one level of nesting out (after line 312)

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs (hunk 1 above)
Comment (alrod) [reply src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs:260]: I think there can be too much noise in the logs

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs (hunk 2 above)
Comment (alrod) [reply src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs:306]: Comment is enhanced

**Comment by alrod** [comment]
> > ```
> >                                 if (_isSessionsEnabled)
> > ```
> 
> Not related to you PR, but why are we not abandoning the messages in case of non-sessions?
> 
> Refers to: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs:293 in [9799ed1](https://github.com/Azure/azure-functions-servicebus-extension/commit/9799ed1b7355c4fd027ff59690932c6d62a147b3). [](commit_id = [9799ed1](https://github.com/Azure/azure-functions-servicebus-extension/commit/9799ed1b7355c4fd027ff59690932c6d62a147b3), deletion_comment = False)

We are relying on lock token expiration in case of non-session. This code was added because delivery count is not increased in case of sessions some time. I think it's also good to manually abandon non-session messages. So I remove `if (_isSessionsEnabled)` as part of this PR.

**Comment by alrod** [comment]
> > ```
> >                             }
> > ```
> 
> If this throws, you will never end up closing and recreating the session. Lets say you enter into SessionLockLost exception once, and then you'd never recovver from that
> 
> Refers to: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs:287 in [9799ed1](https://github.com/Azure/azure-functions-servicebus-extension/commit/9799ed1b7355c4fd027ff59690932c6d62a147b3). [](commit_id = [9799ed1](https://github.com/Azure/azure-functions-servicebus-extension/commit/9799ed1b7355c4fd027ff59690932c6d62a147b3), deletion_comment = False)

From  https://docs.microsoft.com/en-au/azure/service-bus-messaging/message-sessions#session-features:
_The lock is released when Close or CloseAsync are called, or when the lock expires in cases in which the application is unable to do the close operation._
When will the session lock expire? Will IMessageSession close automatically on session lock expire?

I added code to close IMessageSession in catch block.

File: src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs (hunk 4 above)
Comment (alrod) [reply src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs:257]: There will be a NullReferenceException and we will try to accept a new session on the next loop iteration.

**Comment by sidkri** [comment]
Since the default PrefetchCount will remain 0 after this change, I recommend we lower the default value of "MaxMessageCount" to something more likely to be streamed in within the 20ms window. This of course depends on message size but defaulting to a lower value will work for most users and users that want a higher batch size can change the value and when we document that PrefetchCount should be set to the same or higher value, they will hopefully see that avoid running into issues with no prefetch and large batch sizes.  Based on testing so far, a default value around 5 or 10 sounds best.  When we provide a Batch specific PrefetchCount configuration option and default it the same default as MaxMessageCount, we can increase the default values to a higher value.

**Comment by alrod** [comment]
> Since the default PrefetchCount will remain 0 after this change, I recommend we lower the default value of "MaxMessageCount" to something more likely to be streamed in within the 20ms window. This of course depends on message size but defaulting to a lower value will work for most users and users that want a higher batch size can change the value and when we document that PrefetchCount should be set to the same or higher value, they will hopefully see that avoid running into issues with no prefetch and large batch sizes. Based on testing so far, a default value around 5 or 10 sounds best. When we provide a Batch specific PrefetchCount configuration option and default it the same default as MaxMessageCount, we can increase the default values to a higher value.

As discussed we a re not going to change default values.

**Comment by nemakam** [comment]
@alrod, I took a look at the code again, and it looks good to me as well. (y0

**Pull request merged**